### PR TITLE
publish new codec when subscriber needed

### PR DESCRIPTION
**Pull request by cnderrauber**
When publisher publish a track, it can declare a primary codec and some backup codecs, then publish primary codec of the track. When a subscriber joined, found published codecs (only primary at first) not decodable, the sfu server will notify client to publish a matched backup codec for that track. Improve device compatibility

### CODE REVIEW

File: pkg/rtc/mediatrack.go [hunk 1]
```diff
@@ -96,8 +97,7 @@ func (t *MediaTrack) HasSdpCid(cid string) bool {
 		return true
 	}
 
-	info := t.MediaTrackReceiver.TrackInfo(false)
-	t.params.Logger.Debugw("MediaTrack.HasSdpCid", "cid", cid, "trackInfo", info.String())
```
Comment (cnderrauber) [inline pkg/rtc/mediatrack.go:100]: @boks1971 remove log

File: go.sum [hunk 2]
```diff
@@ -135,6 +135,10 @@ github.com/lithammer/shortuuid/v3 v3.0.6 h1:pr15YQyvhiSX/qPxncFtqk+v4xLEpOZObbsY
 github.com/lithammer/shortuuid/v3 v3.0.6/go.mod h1:vMk8ke37EmiewwolSO1NLW8vP4ZaKlRuDIi8tWWmAts=
 github.com/livekit/protocol v0.13.3-0.20220613185908-e66f32ebb7f5 h1:uSOLns5ijvm2z19ppei9iJOYs+1Qf1V6Qrp3M4GcQOc=
 github.com/livekit/protocol v0.13.3-0.20220613185908-e66f32ebb7f5/go.mod h1:BLtSeVmn2rLP37xjzw7gHgaAmkWl3L/L9bPvgSbaOfo=
+github.com/livekit/protocol v0.13.3 h1:MLNeRxuPBHqFjeNlU6q9pu3sYxhQGvM2kZ4eTYHQPTk=
+github.com/livekit/protocol v0.13.3/go.mod h1:BLtSeVmn2rLP37xjzw7gHgaAmkWl3L/L9bPvgSbaOfo=
```
Comment (boks1971) [inline go.sum:139]: `go mod tidy` not cleaning this up?

File: pkg/rtc/mediatrackreceiver.go [hunk 3]
```diff
@@ -151,6 +168,36 @@ func (t *MediaTrackReceiver) SetupReceiver(receiver sfu.TrackReceiver, priority
...
+			t.receivers = append(t.receivers, &simulcastReceiver{
+				TrackReceiver: NewDumbReceiver(livekit.TrackID(t.trackInfo.Sid), string(t.PublisherID()), c, headers),
+				priority:      i,
+			})
+		}
+	}
+	sort.Slice(t.receivers, func(i, j int) bool {
+		return t.receivers[i].Priority() < t.receivers[j].Priority()
+	})
+	t.shadowReceivers()
```
Comment (boks1971) [inline pkg/rtc/mediatrackreceiver.go:192]: Suggest renaming to `shadowReceiversLocked` to follow convention from other places.

File: pkg/rtc/wrappedreceiver.go [hunk 4]
```diff
@@ -47,10 +55,139 @@ func (r *WrappedReceiver) DetermineReceiver(codec webrtc.RTPCodecCapability) {
...
-	for _, receiver := range r.receivers {
-		codecs = append(codecs, receiver.Codec().RTPCodecCapability)
-	}
+func (r *WrappedReceiver) Codecs() []webrtc.RTPCodecParameters {
+	codecs := make([]webrtc.RTPCodecParameters, len(r.receivers))
+	copy(codecs, r.codecs)
 	return codecs
 }
+
+type DumbReceiver struct {
```
Comment (boks1971) [inline pkg/rtc/wrappedreceiver.go:64]: What is the purpose of `DumbReceiver`? Seems like it is keeping state. So, am curious.

**Review by boks1971 — APPROVED** [review]

File: pkg/rtc/wrappedreceiver.go (hunk 4 above)
Comment (cnderrauber) [reply pkg/rtc/wrappedreceiver.go:64]: For a backup codec, we don't have a webrtcReceiver before it published, so create a DumbReceiver to take position of the WebRtcReceiver to negotiate with subscriber downtrack, then we know the actual codec that subscriber needed, feed back to the publisher (through SubscribedCodecQuality), after publisher publish the needed backup codec, DumpReceiver upgrade to webrtcReceiver, everything work as normal

**Pull request merged**